app/controllers/whats_app/webhook_controller.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+module WhatsApp
+  class WebhookController < ApplicationController
+    skip_before_action :require_login, :verify_authenticity_token
+    UNSUCCESSFUL_DELIVERY = %w[undelivered failed].freeze
+
+    def message
+      adapter = WhatsAppAdapter::Inbound.new
+
+      adapter.on(WhatsAppAdapter::UNKNOWN_CONTRIBUTOR) do |whats_app_phone_number|
+        exception = WhatsAppAdapter::UnknownContributorError.new(whats_app_phone_number: whats_app_phone_number)
+        ErrorNotifier.report(exception)
+      end
+
+      adapter.on(WhatsAppAdapter::RESPONDING_TO_TEMPLATE_MESSAGE) do |contributor|
+        contributor.update!(whats_app_message_template_responded_at: Time.current, whats_app_template_message_sent_at: nil)
+        message = contributor.received_messages.first
+        WhatsAppAdapter::Outbound.send!(message)
+      end
+
+      adapter.on(WhatsAppAdapter::UNKNOWN_CONTENT) do |contributor|
+        WhatsAppAdapter::Outbound.send_unknown_content_message!(contributor)
+      end
+
+      whats_app_webhook_params = webhook_params.to_h.transform_keys(&:underscore)
+      adapter.consume(whats_app_webhook_params) { |message| message.contributor.reply(adapter) }
+    end
+
+    def errors
+      return unless webhook_params['Level'] == 'ERROR' && JSON.parse(params['Payload'])['error_code'].to_i == 21_617
+
+      payload = JSON.parse(webhook_params['Payload'])
+      more_info = payload['more_info']
+      parameters = payload['webhook']['request']['parameters']
+      error_text = more_info['Msg'].gsub(Setting.twilio_account_sid, '<ACCOUNT_SID>')
```

Comment:
Since you seem to have the `Setting.twilio_account_sid` in the payload, you could implement a reverse proxy that forwards the errors (http requests, whatever) to the right 100eyes instance. This is a workaround for the fact that we have only one error URL at Twilio.

Reply:
Don't know how to set up a reverse proxy. I'd be interested in learning, but I'm not sure it's worth the time spent right now.

I find it a bit strange that we have this code deployed to all instances and are only able to configurte one webhook endpoint for errors in Twilio, but I'm not sure what benefit we'd gain by having this reverse proxy. If for some reason the instance we configure in Twilio happens to be unreachable when an error happens, we'd likely miss the reporting, but depening on the issue the reverse proxy might also be unavailable :shrug: 

Also, would the idea be that only one instance would have the reverse proxy?

I thought of maybe having a microservice, which exposes an endpoint whose job it is to report the errors for all instances, then the code would not live on any instance, but there are also questions with that approach. Anything other than what we have would take more time, so I think what we have is an acceptable solution for now.